sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java

```diff
@@ -0,0 +1,284 @@
+  public void testInStreamingMode() {
+    TestStream.Builder<KV<String, String>> streamBuilder =
+        TestStream.create(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()))
+            .advanceWatermarkTo(startInstant);
+    for (TimestampedValue<KV<String, String>> element : data) {
+      streamBuilder = streamBuilder.addElements(element);
+    }
+    TestStream<KV<String, String>> stream =
+        streamBuilder
+            .advanceWatermarkTo(startInstant.plus(Duration.standardSeconds(WINDOW_DURATION - 1)))
+            .advanceWatermarkTo(startInstant.plus(Duration.standardSeconds(WINDOW_DURATION + 1)))
+            .advanceWatermarkTo(startInstant.plus(Duration.standardSeconds(NUM_ELEMENTS)))
+            .advanceWatermarkToInfinity();
+
+    PCollection<KV<String, String>> inputCollection =
+        pipeline
+            .apply(stream)
+            .apply(
+                Window.<KV<String, String>>into(
+                    FixedWindows.of(Duration.standardSeconds(WINDOW_DURATION))));
+    inputCollection.apply(
+        ParDo.of(
+            new DoFn<KV<String, String>, Void>() {
+              @ProcessElement
+              public void processElement(ProcessContext c, BoundedWindow window) {
+                LOGGER.debug(
+                    "*** ELEMENT: (%s,%s) *** with timestamp %s in window %s",
+                    c.element().getKey(),
+                    c.element().getValue(),
+                    c.timestamp().toString(),
+                    window.toString());
+              }
+            }));
+
+    // elements have the same key and collection is divided into windows,
+    // so Count.perKey values are the number of elements in windows
+    PCollection<KV<String, Long>> countInput =
+        inputCollection.apply(
+            "Count elements in windows before applying batchingParDo",
+            Count.<String, String>perKey());
+    PAssert.that("Wrong number of elements in windows before BatchingParDo", countInput)
+        .satisfies(new CheckValuesFn(WINDOW_DURATION));
+
+    PCollection<KV<String, String>> outputCollection =
+        inputCollection
+            .apply(batchingParDo)
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+
+    PAssert.that("All elements have not been processed", outputCollection)
+        .satisfies(new CheckAllElementsProcessingFn());
+
+    // elements have the same key and collection is divided into windows,
+    // so Count.perKey values are the number of elements in windows
+    PCollection<KV<String, Long>> countOutput =
+        outputCollection.apply(
+            "Count elements in windows after applying batchingParDo",
+            Count.<String, String>perKey());
+
+    PAssert.that("Wrong number of elements in windows after BatchingParDo", countOutput)
+        .satisfies(new CheckValuesFn(WINDOW_DURATION));
+    pipeline.run().waitUntilFinish();
+  }
+
+  private boolean checkAllElementsProcessing(
```

Comment:
I don't understand what this is actually checking. A comment would be welcome, at minimum.